libs/sdk-ui-dashboard/src/presentation/topBar/buttonBar/button/saveButton/DefaultSaveButton.tsx

```diff
@@ -32,25 +32,33 @@ export function useSaveButtonProps(): ISaveButtonProps {
     const intl = useIntl();
     const emptyTitle = intl.formatMessage({ id: "untitled" });
 
-    const onSaveClick = useCallback(
-        () =>
-            dispatchAndWaitFor(
-                dispatch,
-                // the || is intentional, we want to replace empty string as well
-                saveDashboard(title || emptyTitle),
-            ).then(() => {
-                dispatch(cancelEditRenderMode());
-            }),
-        [dispatch, emptyTitle, title],
-    );
+    // In some cases, when you click the save button (for example immediately after you insert an insight)
+    // and the dashboard is still processing some related requests to this insight, it can take few seconds,
+    // before the CMD.SAVE starts to process (as dashboard can process only 1 command at a time).
+    // We want to show the loading indicator immediately and avoid any double-save race conditions,
+    // so add a local state for this case.
+    const [optimisticIsSaving, setOptimisticIsSaving] = useState(false);
+
+    const onSaveClick = useCallback(() => {
+        setOptimisticIsSaving(true);
+        dispatchAndWaitFor(
+            dispatch,
+            // the || is intentional, we want to replace empty string as well
+            saveDashboard(title || emptyTitle),
+        ).then(() => {
+            setOptimisticIsSaving(false);
+            dispatch(cancelEditRenderMode());
```

Comment:
What about `catch`? We should cancel the optimistic saving there as well, right?

Reply:
I add it, but I think when the save will fail, it wont work anyway :)